### PR TITLE
[feature/build-flags] Add build flags support

### PR DESCRIPTION
**Pull request by felix-schwarz**
## Description
Add support for app build flags to enable/disable features at compile time:
- via `APP_BUILD_FLAGS` in the Xcode project settings
- via `app.build-flags` in the `Branding.plist`:

```xml
<key>app.build-flags</key>
<string>DISABLE_BACKGROUND_LOCATION</string>
```

Overview and instructions in [`doc/APP_BUILD_FLAGS.md`](https://github.com/owncloud/ios-app/blob/feature/build-flags/doc/APP_BUILD_FLAGS.md).

Also includes fixes from `fix/ios12-branded-settings-crash` (#1018).

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by michaelstingl** [comment]
> * via `app.build-flags` in the `Branding.plist`

😍 👍

**Comment by michaelstingl** [comment]
@felix-schwarz in iOS 12 with `DISABLE_BACKGROUND_LOCATION`, "Background upload notifications" are still visible. I think it should be disabled as well.

image

**Comment by felix-schwarz** [comment]
@michaelstingl Thanks! Fixed as of d4e8aa8.

File: ownCloud/AppDelegate.swift [hunk 1]
```diff
@@ -103,8 +103,6 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 		OCExtensionManager.shared.addExtension(UploadMediaAction.actionExtension)
 		OCExtensionManager.shared.addExtension(UploadCameraMediaAction.actionExtension)
 		OCExtensionManager.shared.addExtension(UnshareAction.actionExtension)
-		OCExtensionManager.shared.addExtension(BackgroundFetchUpdateTaskAction.taskExtension)
-		OCExtensionManager.shared.addExtension(InstantMediaUploadTaskExtension.taskExtension)
```
Comment (hosy) [inline ownCloud/AppDelegate.swift:107]: Why are these extensions removed? Looks like a merge conflict!?

File: ownCloud/AppDelegate.swift [hunk 2]
```diff
@@ -134,8 +134,6 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 		OCExtensionManager.shared.addExtension(BackgroundFetchUpdateTaskAction.taskExtension)
 		OCExtensionManager.shared.addExtension(InstantMediaUploadTaskExtension.taskExtension)
 		OCExtensionManager.shared.addExtension(PendingMediaUploadTaskExtension.taskExtension)
-		OCExtensionManager.shared.addExtension(ImportPasteboardAction.actionExtension)
-		OCExtensionManager.shared.addExtension(CutAction.actionExtension)
```
Comment (hosy) [inline ownCloud/AppDelegate.swift:138]: Why are these extensions removed? Looks like a merge conflict!?

**Review by hosy — CHANGES_REQUESTED** [review]

File: ownCloud/AppDelegate.swift (hunk 1 above)
Comment (felix-schwarz) [reply ownCloud/AppDelegate.swift:107]: These extensions were added twice (also in line 134 and 135), so I removed one of the entries.

File: ownCloud/AppDelegate.swift (hunk 2 above)
Comment (felix-schwarz) [reply ownCloud/AppDelegate.swift:138]: I moved these from the "Task extension" section of the code up to the section where the other action extensions are added (now in line 115 + 116).

**Review by hosy — APPROVED** [review]

**Comment by jesmrec** [comment]
Where should the `<key>app.build-flags</key>` be placed in the `Branding.plist` file?? 

i added it in the first `<dict>` block of [oC online file](https://github.com/owncloud/customer-themes-tmp/blob/master/ownCloud.online/iOS/Theming/Branding.plist), and the background options in `Media Uploads` do not disappear.

I guess, this feature is applying only to such parameter, no other parameters ftm. Is that correct?

**Comment by jesmrec** [comment]
Where should be placed the `<key>app.build-flags</key>` inside the `Branding.plist` file?

I set it in the first `<dict>` block of [oC online branding file](https://github.com/owncloud/customer-themes-tmp/blob/master/ownCloud.online/iOS/Theming/Branding.plist) and the background options in `Settings`(Media Upload) are visible and could be switched on.

What's  wrong on my side?

**Comment by hosy** [comment]
@jesmrec this value will only be used, if the app will be build with `fastlane`. You placed the key/value pair on the right place, but you have to build the app via fastlane.

**Comment by hosy** [comment]
@jesmrec should I help you with the fastlane build?

**Comment by jesmrec** [comment]
Thanks @hosy for your help. Just as info.

To build the app:  `fastlane owncloud_branding_adhoc_build` , by using correct distribution cert.

With the branding parameter

```
<key>app.build-flags</key>
<string>DISABLE_BACKGROUND_LOCATION</string>
```

the background location services are disabled:

Screenshot 2021-09-17 at 14 23 42

And the ´Info.plist` does not contain the `DISABLE_BACKGROUND_LOCATION` stuff

**Comment by jesmrec** [comment]
So, this is approved on my side

**Review by jesmrec — APPROVED** [review]

**Pull request merged**